File: dart/question14.dart
void main() {}

String longestCommonPrefix(List<String> strs) {
  var prefix = strs[0];

  while (prefix.length != 0) {
    for (final singleString in strs) {
      if (startWithPrefix(strs, prefix)) {
        return prefix;
      } else {
        prefix = prefix.substring(0, prefix.length - 1);
      }
    }
  }
  return prefix;
}

bool startWithPrefix(List<String> strs, String prefix) {
  for (final everyString in strs) {
    final boo = everyString.startsWith(prefix);
    if (!boo) {
      return false;
    }
  }
  return true;
}

File: src/question14.rs

pub fn longest_common_prefix(strs: Vec<String>) -> String {
    let mut current_prefix: &str = &strs[0];

    while !current_prefix.is_empty() {
        if all_start_with(current_prefix, &strs) {
            return current_prefix.to_string();
        } else {
            current_prefix = &current_prefix[0..current_prefix.len() - 1];
        }
    }
    current_prefix.to_string()
}

pub fn all_start_with(prefix: &str, strs: &[String]) -> bool {
    for single_string in strs {
        let boo = single_string.starts_with(prefix);
        if !boo {
            return false;
        }
    }
    true
}